src/charm.py

```diff
@@ -950,20 +945,48 @@ def _create_or_update_kv_secret(
         role_name: str,
         role_id: str,
         role_secret_id: str,
-        id: Optional[str] = None,
     ) -> Secret:
         """Create or update the KV secret for the relation.
 
         Args:
             role_name: The role name to set the secret for
             role_id: The role ID to set in the secret
             role_secret_id: The role secret ID to set in the secret
-            id: Secret ID that should be updated
         """
         juju_secret_label = f"{KV_SECRET_PREFIX}{role_name}"
-        return self._set_juju_secret(
-            juju_secret_label, {"role-id": role_id, "role-secret-id": role_secret_id}, id=id
+        # TODO bug: https://bugs.launchpad.net/juju/+bug/2075153
+        # Until the reference bug is fixed we must pass the secret ID here
+        # not to lose the secret://modeluuid:secretID format
+        secret_id_from_peer_relation_data = self._get_vault_kv_secrets_in_peer_relation().get(
+            juju_secret_label
+        )
+        secret = self._set_juju_secret(
+            juju_secret_label,
+            {"role-id": role_id, "role-secret-id": role_secret_id},
+            id=secret_id_from_peer_relation_data,
         )
+        if secret.id is None:
+            raise RuntimeError(f"Unexpected error, created secret {juju_secret_label!r} has no id")
+        if secret.id != secret_id_from_peer_relation_data:
+            self._set_vault_kv_secret_in_peer_relation(juju_secret_label, secret.id)
+        return secret
+
+    def _set_vault_kv_secret_in_peer_relation(self, label: str, secret_id: str):
+        """Set the vault kv secret in the peer relation."""
+        if not self._is_peer_relation_created():
+            raise RuntimeError("Peer relation not created")
+        secrets = self._get_vault_kv_secrets_in_peer_relation()
+        secrets[label] = secret_id
+        relation = self.model.get_relation(PEER_RELATION_NAME)
+        relation.data[self.app].update({"vault-kv-secrets": json.dumps(secrets, sort_keys=True)})  # type: ignore[union-attr]  # noqa: E501
```

Comment:
I'm not a fan of this, there is no mechanism to clean these up, can't we leverage the `secret-ids` call instead?

Reply:
No, currently that won't work either.
`secret-ids` gives us a list of secret IDs as its name suggests, then we can use that id in `get-secret` or in `secret-info-get` to get the content and the label of the secret.
Looking at the ID alone we won't be able to decide which secret should be deleted unless we keep track of the IDs so we know to which requirer it belongs.
Then looking at the info of the secret for example looking at the label we go back to square one where we need to know the label to delete the secret.
And looking at the content of the secret won't help either because that content can't be calculated again in association with the requirer app to do the comparison.

In the bug reported here where using `secret-ids` was suggested it could make sense (haven't fully validated it yet) because we would look at the content of the secret where we see a csr that we can compare for example to what we have in relation data (although that can be troublesome too if we lose access to relation data. Needs further investigation).

I see only two ways forward here:
1. The solution proposed in this PR (btw used in k8s too)
2. Refactoring Vault-kv so the provider and the requirer can both compute the same secret label without the need for information that we might not have if not passed by the other party.

Reply:
I wasn't entirely right in the previous comment.
Iterating over the result of `secret-ids` and getting the label of each secret using `secret-info-get` we can delete the secret by knowing part of the label not the whole label. For example in Vault KV it would be enough to find the unit name in the secret label.
But I would not see that solution as ideal as we'd need to iterate over all secrets and their info when removing any requirer relation.

I think this an interesting discussion of how to clean up unused secrets.

Reply:
So this extends the list of possible solutions from the first comment to 3:
1. The solution proposed in this PR (btw used in k8s too)
2. Refactoring Vault-kv so the provider and the requirer can both compute the same secret label without the need for information that we might not have if not passed by the other party.
3. Iterate over secrets by calling `secret-info-get` using the IDs returned by `secret-ids` to delete the secret based on partial label match.

Reply:
Let me know if you prefer to take this into a team discussion or if you have a strong opinion on which solution seems best from the previous list. I am strongly in favour of .2 and the refactor of Vault KV.

Reply:
I'm not sure I fully understand option 2, but I would fight back against any option (like .1) where we store data that can be fetched by another mean. Then we need to make sure this data is in sync with the real source of truth.

Reply:
Option 2 is the one I went for and implemented in #266 
Closing this PR in favour of that one.